### PR TITLE
feature/#271_Allow_admins_to_customize_the_Profile_screen_for_users 

### PR DESCRIPTION
**Pull request by olatechpro**
- Allow admins to customize the "Profile" screen for users #271

profile-features

@agapetry @stevejburge Here are few things about this PR.

### Element update method
- I added 2 buttons with first one redirect admin to the current role profile edit screen to enable us retrieve all the profile fields and element for update while the second one redirect to the current user profile fields since profile fields items may differ by roles, users or when editing own profile to when editing others profile
- Some plugins may add element without class or id and as a result, we're automatically generating new class for elements without class

### Profile features UI
- The UI are only organised to make it easier to browse through section and check fields under it but there's no really any relationship between section(title) and fields due to the way the content is organised on profile fields. I only make header element type stand as main element since the fields will always be under it and in order they appear on page.
- The fields order may not be 100% accurate as newly added elements will always come last and as a result, i added a drag and drop features to enable custom sorting if user need better organisation

The changes are merely prototype since the whole concept came from only me and i'm opened to suggestion on improvements.

### CODE REVIEW

**Comment by agapetry** [comment]
@olatechpro Sorry for the delay on this. I think it's a sensible addition. I would like to take some time tomorrow to install it and think over the "element update method." I understand the need to request the page so we can capture the ID and title of fields added by third parties.  The links are sensible, but may be confusing to users. What do you think about redirecting to the Edit Profile screen the first time a role is selected, then back to Profile Features restrictions?

Then instead of the dedicated links always there, you could implement a "refresh fields" link that would:

1. Redirect to Edit Profile, then return to Profile Features Restrictions, for the currently selected role
2. Clear the "done" flags for all the other roles so the same thing happens with them on next selection

**Comment by agapetry** [comment]
> Okay @agapetry I'll make the update and notify you but I think the first redirection need to be 2 as if we redirect to current selected role alone, items that appear when user is editing their own profile will be empty or how do you think we should handle this?

If we are specifying how users of each role have elements hidden on their own Edit Profile screen, it seems there would need to be an implicit user login switch to that role (if not the same as editing user), then a redirect to Edit Profile, then back. I don't follow why it matters whether the editing user sees their own profile fields listed when configuring another role.  Or have I misunderstood your aim?

**Comment by agapetry** [comment]
> If we're capturing elements present on the screen when editing a user, that won't include fields or elements that only show when user are editing their own profile

Right, that's why I suggested switching the login to a user of that role prior to the redirect.

**Comment by agapetry** [comment]
@olatechpro Are we in agreement that this is the functional goal?

> control how users of the selected role have elements hidden on their own Edit Profile screen

I don't see why we would want the Profile Feature Restrictions screen to display items that role doesn't have, since we are only restricting, not adding fields.

**Comment by agapetry** [comment]
@olatechpro Okay, I installed this PR and studied the code. The basic concept is good, but I think you've taken a wrong turn in dealing with "Edit User" screen elements instead of focusing exclusively on Profile elements. The function of this screen should be to restrict what a user sees on their own Profile screen, not what another user sees when editing their user account.

If there is also a desire to control what fields a certain role sees on the Edit User, that should be handled somewhere other than "Profile Restrictions" to avoid confusion. I think "User Editing Restrictions" would have a smaller appeal since on most sites, user editing is by Administrators only.

Here are some details on how I see these factors playing out in the current PR:

1. The user_id argument passed with user-edit.php?ppc_profile_element=1 causes a role-specific Edit User screen to be loaded, but the detected elements for any role are updated to (and later retrieved from) the same 'capsman_profile_features_elements' option array without any role key. So any role-specific differences in available elements on the Edit User screen are lost.  But again, restricting the "Edit User" screen is not our aim anyway. If anything, this could be fixed and applied to a "User Editing Restrictions" screen.

2. Even if we did decide to finish implementing the restriction of Edit User fields, role-specific differences in elements on the Edit Profile screen would not be handled by the current PR. The "update general profile items" request will cause the Administrator's profile screen elements to be added as available Profile Restriction elements for all roles. Here are some examples of items that do not apply to the Subscriber role, but would be displayed as an available Profile Restriction for that role:

- Personal Options > Syntax Highlighting
- WPML > Editing Language
- WPML > HIdden Languages
- Custom User Permissions
- Permission Groups
- Additional Capabilities
- Yoast SEO Settings
- Forums > Forum Role
- Editorial Notifications (PublishPress Planner)

This is the result from six familiar or commonly used plugins, but there would be no end to examples of other plugin-generated fields which are specific to the Administrator or Editor role.

3. Whatever implementation is used to trigger an elements refresh, it would be better to do that automatically the first time a role is accessed for restriction settings, rather than opening with a blank screen that requires the user to discover and follow a "click to update" link. If we can agree to focus exclusively on Profile element restriction, this will become a single update button / link. 

4. If we do an elements auto-refresh for first time Profile Restrictions role selections, further automation could be achieved by also clearing the "done" flag on new plugin install (forcing another auto-refresh on next Profile Restrictions visit).

In summary, Profile Restrictions should apply to the Profile screen only, not Edit User. To avoid confusion, those Profile Restrictions should not present any checkboxes for elements that don't apply to the selected role.  I don't see how to achieve that without a user switch redirect, but please let me know if I'm still misunderstanding anything.

**Comment by stevejburge** [comment]
@agapetry Thanks for the detailed feedback.

I'm OK with this. Our goal with Capabilities is to control what ordinary users can do, not Administrators.

So it's OK if we skip the "Edit User" screen, and certainly in the first version

**Comment by agapetry** [comment]
@olatechpro The user switch redirect is the only solution I could think of to capture the fields, but I'm open to other suggestions. Otherwise, yes to point 1.

Yes to point 2: clearing the "refresh done" flag for each role so a user redirect will occur next time that role is selected in Profile Restrictions.

For point 3, the manual refresh button would also clear all the "refresh done" flags.  Then a single redirect would occur the first time any role is selected in Profile Restrictions.

**Comment by agapetry** [comment]
@olatechpro It looks like the redirect will need to happen earlier. This is what I see on first execution of the updated PR:

image

**Comment by agapetry** [comment]
@olatechpro Good, I see the redirect refresh working or almost any role I select. The only remaining failure I see is on Administrator role selection:

image

**Comment by agapetry** [comment]
It seems like for auto-switch logic should first use the current user if the selected role matches.

**Comment by agapetry** [comment]
The normal case is for a site to have only one Administrator.

**Comment by agapetry** [comment]
@olatechpro Good, that works now. The only quirk I see is this unexpected message after the Administrator refresh:

image

That could be confusing, making the user wonder what "element" was updated before they selected anything or clicked Update.

**Comment by agapetry** [comment]
@olatechpro What about the "You do not have permission" scenario those messages were designed for?

**Review by agapetry — APPROVED** [review]
